src/overlay/PendingEnvelopes.cpp

```diff
@@ -0,0 +1,213 @@
+    if (items.empty())
+    {
+        return EnvelopeHandler::ENVELOPE_STATUS_READY;
+    }
+    else
+    {
+        mEnvelopes[envelope.statement.slotIndex].mFetchingEnvelopes.insert(
+            envelope);
+        mEnvelopeItemMap.add(envelope, items);
+        return EnvelopeHandler::ENVELOPE_STATUS_FETCHING;
+    }
+}
+
+std::pair<bool, std::set<SCPEnvelope>>
+PendingEnvelopes::handleQuorumSet(SCPQuorumSet const& qSet, bool force)
+{
+    auto addResult = mApp.getItemFetcher().add(qSet, force);
+    if (!addResult.first)
+    {
+        if (!isQuorumSetSane(qSet, false))
+        {
+            discardEnvelopesWithItem(addResult.second);
+        }
+        return std::make_pair(false, std::set<SCPEnvelope>{});
+    }
+
+    return std::make_pair(true, processReadyItems(addResult.second));
+}
+
+std::set<SCPEnvelope>
+PendingEnvelopes::handleTxSet(TransactionSet const& txSet, bool force)
+{
+    auto addResult = mApp.getItemFetcher().add(txSet, force);
+    if (!addResult.first)
+    {
+        return std::set<SCPEnvelope>{};
+    }
+
+    return processReadyItems(addResult.second);
+}
+
+std::set<SCPEnvelope>
+PendingEnvelopes::processReadyItems(ItemKey itemKey)
+{
+    auto envelopes = mEnvelopeItemMap.remove(itemKey);
+    for (auto& e : envelopes)
+    {
+        mEnvelopes[e.statement.slotIndex].mFetchingEnvelopes.erase(e);
+    }
+    return envelopes;
+}
+
+void
+PendingEnvelopes::discardEnvelope(SCPEnvelope const& envelope)
+{
+    if (isDiscarded(envelope))
+    {
+        return;
+    }
+
+    mEnvelopes[envelope.statement.slotIndex].mFetchingEnvelopes.erase(envelope);
+    mEnvelopes[envelope.statement.slotIndex].mDiscardedEnvelopes.insert(
+        envelope);
+    for (auto& item : mEnvelopeItemMap.remove(envelope))
+    {
+        mApp.getItemFetcher().forget(item);
+    }
+}
+
+void
+PendingEnvelopes::discardEnvelopesWithItem(ItemKey itemKey)
+{
+    CLOG(TRACE, "Herder") << "Discarding SCP Envelopes with SCPQSet "
+                          << hexAbbrev(itemKey.getHash());
+
+    for (auto& envelope : mEnvelopeItemMap.envelopes(itemKey))
+    {
+        discardEnvelope(envelope);
+    }
+}
+
+void
+PendingEnvelopes::touchItemCache(SCPEnvelope const& envelope)
+{
+    mApp.getItemFetcher().touch(
+        ItemKey{ItemType::QUORUM_SET, getQuorumSetHash(envelope)});
+
+    for (auto const& h : getTxSetHashes(envelope))
+    {
+        mApp.getItemFetcher().touch(ItemKey{ItemType::TX_SET, h});
+    }
+}
+
+bool
+PendingEnvelopes::isDiscarded(SCPEnvelope const& envelope)
+{
+    if (envelope.statement.slotIndex < mMinimumSlotIndex)
+    {
+        return true;
+    }
+
+    auto& discardedSet =
+        mEnvelopes[envelope.statement.slotIndex].mDiscardedEnvelopes;
+    auto discarded =
+        std::find(std::begin(discardedSet), std::end(discardedSet), envelope);
+    return discarded != std::end(discardedSet);
+}
+
+void
+PendingEnvelopes::setMinimumSlotIndex(uint64_t slotIndex)
```

Comment:
we should not be tracking the "minimum slot index" here: this is the responsibility of `OverlayEnvelopeHandler`